File: leetcode/1011.capacity-to-ship-packages-within-d-days.py
# @leet start
class Solution:
    def shipWithinDays(self, weights: list[int], days: int) -> int:
        def is_valid(max_weight: int) -> bool:
            count = 1
            curr = 0

            for weight in weights:
                if curr + weight > max_weight:
                    count += 1
                    curr = weight
                else:
                    curr += weight

            return count <= days

        low = max(weights)
        high = sum(weights)

        while low <= high:
            mid = (low + high) // 2

            if is_valid(mid):
                high = mid - 1
            else:
                low = mid + 1

        return low


# @leet end

File: leetcode/1011.capacity-to-ship-packages-within-d-days.js
// @leet start
/**
 * @param {number[]} weights
 * @param {number} days
 * @return {number}
 */
var shipWithinDays = function (weights, days) {
  const isValid = (capacity) => {
    let extraDays = days - 1;
    let currCapacity = capacity;

    for (const weight of weights) {
      if (weight > capacity) {
        return false;
      }

      if (weight > currCapacity) {
        if (extraDays === 0) {
          return false;
        }

        extraDays--;
        currCapacity = capacity;
      }

      currCapacity -= weight;
    }

    return true;
  };

  let left = Math.max(...weights);
  let right = weights.reduce((acc, w) => acc + w);

  while (left <= right) {
    const mid = Math.floor((left + right) / 2);

    if (isValid(mid)) {
      right = mid - 1;
    } else {
      left = mid + 1;
    }
  }

  return left;
};
// @leet end
